### PR TITLE
v2 plugins: add catkin plugin

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

Add a new Catkin plugin using v2 of the plugin API. Refactor common functions out of the Colcon plugin to share functionality with both.

### CODE REVIEW

**Comment by kyrofa** [comment]
> an additional spread test to exercise all of the options

Certainly, I'll add that tomorrow.

**Comment by kyrofa** [comment]
Alright, I pushed a few more spread tests to fully cover the plugin options. I'll do one more to ensure launching works properly then this will be ready.

File: tests/spread/plugins/v2/catkin-ros1-catkin-packages-ignore/task.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+execute: |
+  cd "../snaps/$SNAP"
+
+  # Build the snap and verify that only package2 was included
+  snapcraft
+  [ -d parts/test-part/install/lib/package2 ]
+  [ ! -d parts/test-part/install/lib/package1 ]
+
+  snap install "${SNAP}"_1.0_*.snap --dangerous
+  [ "$($SNAP)" = "I am package2" ]
```
Comment (cjp256) [inline tests/spread/plugins/v2/catkin-ros1-catkin-packages-ignore/task.yaml:39]: Couple of these files have no trailing newline for whatever reason..

Comment (kyrofa) [reply tests/spread/plugins/v2/catkin-ros1-catkin-packages-ignore/task.yaml:39]: Are you saying you want them?

**Comment by kyrofa** [comment]
Darn we're running out of time here on spread now. Can I increase the runners?

File: tests/spread/plugins/v2/catkin-ros1-catkin-packages-ignore/task.yaml (hunk 1 above)
Comment (cjp256) [reply tests/spread/plugins/v2/catkin-ros1-catkin-packages-ignore/task.yaml:39]: Doesn't matter, just a nit.

**Review by cjp256 — APPROVED** [review]
Looks great!  I really like what you did with _ros and the test coverage on this round. Thank you for doing it :+1:

File: snapcraft/internal/project_loader/_extensions/ros1_noetic.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+from typing import Any, Dict, Optional, Tuple
+
+from typing_extensions import Final
+
+from ._extension import Extension
+
+
+class ExtensionImpl(Extension):
+    """Drives ROS 1 build and runtime environment for snap."""
```
Comment (sergiusens) [inline snapcraft/internal/project_loader/_extensions/ros1_noetic.py:27]: Cosmetic to follow https://www.python.org/dev/peps/pep-0257/

```suggestion
    """Setup a ROS 1 build and runtime environment suitable for snaps."""
```

This part in the text
> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

You can use my suggestion or write a new string based on that

File: snapcraft/internal/project_loader/_extensions/ros2_foxy.py [hunk 3]
```diff
@@ -26,7 +26,7 @@
 
 
 class ExtensionImpl(Extension):
-    """Drives ROS2 build and runtime environment for snap."""
+    """Drives ROS 2 build and runtime environment for snap."""
```
Comment (sergiusens) [inline snapcraft/internal/project_loader/_extensions/ros2_foxy.py:29]: ah, I see how that comment got in (for what it's worth, this is a general issue we have)

File: snapcraft/plugins/v2/_ros.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+import click
+from catkin_pkg import packages as catkin_packages
+
+from snapcraft.internal.repo import Repo
+from snapcraft.plugins.v1._ros.rosdep import _parse_rosdep_resolve_dependencies
+from snapcraft.plugins.v2 import PluginV2
+
+
+class RosPlugin(PluginV2):
```
Comment (sergiusens) [inline snapcraft/plugins/v2/_ros.py:32]: if this is not intended by base users and only meant to be inherited from, how about making it an ABC?

File: snapcraft/plugins/v2/_ros.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {
+            "ROS_PYTHON_VERSION": "3",
+        }
+
+    def _get_workspace_activation_commands(self) -> List[str]:
+        """
+        Return a list of commands source a ROS workspace.
```
Comment (sergiusens) [inline snapcraft/plugins/v2/_ros.py:50]: This line should be next to """ (https://www.python.org/dev/peps/pep-0257/)
```suggestion
        """Return a list of commands source a ROS workspace.
```

File: snapcraft/plugins/v2/_ros.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        snapcraftctl can be used in the script to call out to snapcraft
+        specific functionality.
+        """
+
+        return [". /opt/ros/$ROS_DISTRO/setup.sh"]
+
+    @abc.abstractmethod
+    def _get_build_commands(self) -> List[str]:
+        """
+        Return a list of commands to run during the build step.
```
Comment (sergiusens) [inline snapcraft/plugins/v2/_ros.py:66]:
```suggestion
        """Return a list of commands to run during the build step.
```

File: snapcraft/plugins/v2/catkin.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      specified or set to an empty list ([]), no packages will be ignored.
+
+    - catkin-cmake-args:
+      (list of strings)
+      Arguments to pass to cmake projects. Note that any arguments here which match
+      catkin arguments need to be prefixed with a space. This can be done by quoting
+      each argument with a leading space.
+
+    This plugin expects the build-environment `ROS_DISTRO` and `ROS_BUILD_BASE`
+    to be populated by the `ros-<distro>` extension.
```
Comment (sergiusens) [inline snapcraft/plugins/v2/catkin.py:37]: more that expects, I would say relies on those variables and they can be handled gracefully by the ros-<distro> extension or something along those lines.

Could be like browser support from the Netscape days: "works best with the ros-<distro> extension" :-)

File: snapcraft/plugins/v2/catkin.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    This plugin also expects certain variables that are specified by the extension,
+    specific to the ROS distro.  If not using the extension, set these in your
+    `build-environment`:
+      - ROS_DISTRO: "noetic"
```
Comment (sergiusens) [inline snapcraft/plugins/v2/catkin.py:42]: This feels a bit repetitive when compared with the last paragraph

File: tests/spread/plugins/v2/catkin-ros1-catkin-cmake-args/task.yaml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  #shellcheck source=tests/spread/tools/snapcraft-yaml.sh
+  . "$TOOLS_DIR/snapcraft-yaml.sh"
+  restore_yaml "snap/snapcraft.yaml"
+
+execute: |
+  cd "../snaps/$SNAP"
+  snapcraft
+
+  snap install "${SNAP}"_1.0_*.snap --dangerous
+  [ "$($SNAP)" = "hello snap" ]
```
Comment (sergiusens) [inline tests/spread/plugins/v2/catkin-ros1-catkin-cmake-args/task.yaml:35]:
```suggestion
  [ "$($SNAP)" = "hello snap" ]

```

Comment (sergiusens) [reply tests/spread/plugins/v2/catkin-ros1-catkin-cmake-args/task.yaml:35]: same for others?

**Review by sergiusens — COMMENTED** [review]
Looks really good, I have mostly cosmetic comments

File: snapcraft/plugins/v2/_ros.py (hunk 4 above)
Comment (kyrofa) [reply snapcraft/plugins/v2/_ros.py:32]: Oh it already is. `PluginV2` is an ABC, and this class even defines more abstract methods.

File: snapcraft/internal/project_loader/_extensions/ros2_foxy.py (hunk 3 above)
Comment (kyrofa) [reply snapcraft/internal/project_loader/_extensions/ros2_foxy.py:29]: Yeah it happens. I won't fix this one here, but no need to continue duplicating a bad pattern.

File: snapcraft/internal/project_loader/_extensions/ros1_noetic.py (hunk 2 above)
Comment (kyrofa) [reply snapcraft/internal/project_loader/_extensions/ros1_noetic.py]: Fixed.

File: snapcraft/plugins/v2/_ros.py (hunk 5 above)
Comment (kyrofa) [reply snapcraft/plugins/v2/_ros.py]: Fixed.

File: snapcraft/plugins/v2/_ros.py (hunk 6 above)
Comment (kyrofa) [reply snapcraft/plugins/v2/_ros.py]: Fixed.

File: snapcraft/plugins/v2/catkin.py (hunk 7 above)
Comment (kyrofa) [reply snapcraft/plugins/v2/catkin.py]: Indeed, rephrased.

File: snapcraft/plugins/v2/catkin.py (hunk 8 above)
Comment (kyrofa) [reply snapcraft/plugins/v2/catkin.py:37]: Agreed, I have rephrased.

File: tests/spread/plugins/v2/catkin-ros1-catkin-cmake-args/task.yaml (hunk 9 above)
Comment (kyrofa) [reply tests/spread/plugins/v2/catkin-ros1-catkin-cmake-args/task.yaml]: Sure, done.

File: snapcraft/plugins/v2/_ros.py (hunk 4 above)
Comment (sergiusens) [reply snapcraft/plugins/v2/_ros.py:32]: Nice! Missed that part

**Review by sergiusens — APPROVED** [review]

**Pull request merged**